src/rnnlm/arpa-sampling.cc

```diff
@@ -0,0 +1,145 @@
+// arpa-sampling.cc
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <math.h>
+#include <string>
+
+#include "arpa-sampling.h"
+
+namespace kaldi {
+
+// This function reads in each ngram line from an ARPA file
+void ArpaSampling::ConsumeNGram(const NGram& ngram) {
+  int32 cur_order = ngram.words.size();
+  int32 word = ngram.words.back(); // word is the last word in a ngram term 
+  HistType history(ngram.words.begin(), ngram.words.begin() + cur_order - 1);
+  KALDI_ASSERT(history.size() == cur_order - 1);
+
+  // log probability of a ngram term (natural log)
+  BaseFloat log_prob = ngram.logprob;
+  // backoff log probability of a ngram term (natural log)
+  BaseFloat backoff_weight = ngram.backoff;
+  std::pair <BaseFloat, BaseFloat> probs_pair;
+  probs_pair = std::make_pair(log_prob, backoff_weight);
+  probs_[cur_order - 1][history].insert({word, probs_pair});
+ 
+  // get the total number of words from an ARPA file
+  if (cur_order == 1) {
+    num_words_++;
+  }
+}
+
+void ArpaSampling::HeaderAvailable() {
+  ngram_counts_ = NgramCounts();
+  ngram_order_ = NgramCounts().size(); 
+  probs_.resize(ngram_order_);
+}
+
+BaseFloat ArpaSampling::GetProb(int32 order, int32 word, const HistType &history) {
+  BaseFloat prob = 0.0;
+  NgramType::const_iterator it = probs_[order - 1].find(history);
+  if (it != probs_[order - 1].end() &&
+      probs_[order-1][history].find(word) != probs_[order-1][history].end()) {
+    prob += probs_[order-1][history][word].first;
+  } else { // backoff to the previous order
+    order--;
```

Comment:
Best not to change 'order', instead refer hereafter to 'order - 1'.  Modifying variables
leads to code that's more bug-prone in general.